src/verify_cert.rs

```diff
@@ -341,42 +333,68 @@ fn check_basic_constraints(
     }
 }
 
+/// Extended Key Usage (EKU) of a certificate.
+pub enum ExtendedKeyUsage {
+    /// The certificate must contain the specified [`KeyPurposeId`] as EKU.
+    Required(KeyPurposeId),
+
+    /// If the certificate has EKUs, then the specified [`KeyPurposeId`] must be included.
+    RequiredIfPresent(KeyPurposeId),
+}
+
+impl ExtendedKeyUsage {
+    /// Get the [`KeyPurposeId`] of either variant.
+    fn key_purpose_id(&self) -> KeyPurposeId {
+        match self {
+            ExtendedKeyUsage::Required(eku) => *eku,
+            ExtendedKeyUsage::RequiredIfPresent(eku) => *eku,
+        }
+    }
+}
+
+/// An OID value indicating an Extended Key Usage (EKU) key purpose.
 #[derive(Clone, Copy, PartialEq, Eq)]
-pub(crate) struct KeyPurposeId {
+pub struct KeyPurposeId {
     oid_value: untrusted::Input<'static>,
 }
 
+impl KeyPurposeId {
+    /// Construct a new [`KeyPurposeId`].
+    ///
+    /// `oid` is the OBJECT IDENTIFIER in bytes.
+    pub const fn new(oid: &'static [u8]) -> Self {
+        KeyPurposeId {
+            oid_value: untrusted::Input::from(oid),
+        }
+    }
+}
+
 // id-pkix            OBJECT IDENTIFIER ::= { 1 3 6 1 5 5 7 }
 // id-kp              OBJECT IDENTIFIER ::= { id-pkix 3 }
 
 // id-kp-serverAuth   OBJECT IDENTIFIER ::= { id-kp 1 }
 #[allow(clippy::identity_op)] // TODO: Make this clearer
-pub(crate) static EKU_SERVER_AUTH: KeyPurposeId = KeyPurposeId {
-    oid_value: untrusted::Input::from(&[(40 * 1) + 3, 6, 1, 5, 5, 7, 3, 1]),
-};
+pub(crate) static EKU_SERVER_AUTH: KeyPurposeId =
+    KeyPurposeId::new(&[(40 * 1) + 3, 6, 1, 5, 5, 7, 3, 1]);
 
 // id-kp-clientAuth   OBJECT IDENTIFIER ::= { id-kp 2 }
 #[allow(clippy::identity_op)] // TODO: Make this clearer
-pub(crate) static EKU_CLIENT_AUTH: KeyPurposeId = KeyPurposeId {
-    oid_value: untrusted::Input::from(&[(40 * 1) + 3, 6, 1, 5, 5, 7, 3, 2]),
-};
+pub(crate) static EKU_CLIENT_AUTH: KeyPurposeId =
+    KeyPurposeId::new(&[(40 * 1) + 3, 6, 1, 5, 5, 7, 3, 2]);
 
 // id-kp-OCSPSigning  OBJECT IDENTIFIER ::= { id-kp 9 }
 #[allow(clippy::identity_op)] // TODO: Make this clearer
-pub(crate) static EKU_OCSP_SIGNING: KeyPurposeId = KeyPurposeId {
-    oid_value: untrusted::Input::from(&[(40 * 1) + 3, 6, 1, 5, 5, 7, 3, 9]),
-};
+pub(crate) static EKU_OCSP_SIGNING: KeyPurposeId =
+    KeyPurposeId::new(&[(40 * 1) + 3, 6, 1, 5, 5, 7, 3, 9]);
 
 // https://tools.ietf.org/html/rfc5280#section-4.2.1.12
-fn check_eku(
-    input: Option<&mut untrusted::Reader>,
-    required_eku_if_present: KeyPurposeId,
-) -> Result<(), Error> {
+fn check_eku(input: Option<&mut untrusted::Reader>, eku: &ExtendedKeyUsage) -> Result<(), Error> {
+    let key_purpose = eku.key_purpose_id();
     match input {
         Some(input) => {
             loop {
                 let value = der::expect_tag_and_get_value(input, der::Tag::OID)?;
-                if value == required_eku_if_present.oid_value {
+                if value == key_purpose.oid_value {
```

Comment:
I feel like it might be nicer to move this comparison into a method on the new enum? That way it could probably abstract better over the difference of the present vs not variants, and we don't need the the key_purpose_id() method which feels like a bit of a layering violation.

Reply:
Oh, seems I misunderstood @djc's suggestion, apologies. I thought the point was to not have to use the `key_purpose_id()` in that particular place, as opposed to it existing at all.

There is one other place in [`build_chain_inner()`](https://github.com/rustls/webpki/blob/92fd1ba94d443c62c036e176ac50b440d1db5f47/src/verify_cert.rs#L63) where it is used. I can use the same comparison method there, and then `key_purpose_id()` can go away. If that is what you meant then I'll do that.

Reply:
> If that is what you meant then I'll do that.

Yeah I think that's a good approach. Thanks!

Reply:
Alright, done.